qbit-http-storages/src/commonMain/kotlin/qbit.storage/YandexDiskStorage.kt

```diff
@@ -0,0 +1,250 @@
+
+        val client = HttpClient()
+        val response = getUrlToDownload(client, filePath)
+        val json = Json.indented.parseJson(response.readText())
+        val downloadUrl = (json as JsonObject).getPrimitive("href").content
+
+        val byteArray = getFile(client, downloadUrl)
+        client.close()
+        byteArray
+    }
+
+    @InternalCoroutinesApi
+    override fun keys(namespace: Namespace): Collection<Key> = runBlocking {
+        var fullPath = getFullPath(namespace)
+
+        val client = HttpClient()
+        val response = client.get<HttpResponse>(YANDEX_DISK_API_RESOURCES) {
+            parameter("path", fullPath)
+            header("Authorization", "OAuth $accessToken")
+        }
+        val json = Json(JsonConfiguration.Stable)
+        val resource = json.parse(Resource.serializer(), response.readText())
+        val fileNames = getFileNamesInResourceByType(resource, "file")
+        val keys = wrapFileNamesToKeys(fileNames, namespace)
+        client.close()
+        keys
+    }
+
+    @InternalCoroutinesApi
+    override fun subNamespaces(namespace: Namespace): Collection<Namespace> = runBlocking {
+        var fullPath = getFullPath(namespace)
+
+        val client = HttpClient()
+        val response = client.get<HttpResponse>(YANDEX_DISK_API_RESOURCES) {
+            parameter("path", fullPath)
+            header("Authorization", "OAuth $accessToken")
+        }
+        val json = Json(JsonConfiguration.Stable)
+        val resource = json.parse(Resource.serializer(), response.readText())
+        val dirNames = getFileNamesInResourceByType(resource, "dir")
+        val namespaces = wrapDirNamesToNamespaces(dirNames, namespace)
+        client.close()
+        namespaces
+    }
+
+    @InternalCoroutinesApi
+    override fun hasKey(key: Key): Boolean = runBlocking {
+        var fullPath = getFullPath(key)
+
+        val client = HttpClient()
+        try {
+            val response = client.get<String>(YANDEX_DISK_API_RESOURCES) {
+                parameter("path", fullPath)
+                header("Authorization", "OAuth $accessToken")
+            }
+            client.close()
+            true
+        } catch (e: ClientRequestException) {
+            client.close()
+            false
+        }
+    }
+
+    private suspend fun uploadFile(client: HttpClient, urlToUploadFile: String, byteArray: ByteArray): HttpResponse {
+        return client.put(urlToUploadFile) {
+            body = ByteArrayContent(byteArray)
+        };
+    }
+
+    private suspend fun getUrlToUploadFile(client: HttpClient, filePath: String, fileName: String): HttpResponse {
+        return client.get(YANDEX_DISK_API_GET_FILE_URL) {
+            parameter("path", filePath + fileName)
+            parameter("overwrite", true)
+            header("Authorization", "OAuth $accessToken")
+        }
+    }
+
+    private suspend fun createFolderStructure(client: HttpClient, paths: List<String>) {
```

Comment:
What is the reason to create each folder?

Reply:
if complicated path like "/a/b/c/file.png" doesn't exist, we need to create "a", then "b" inside c, and etc. Yandex Disk api doesn't support single request for creating the whole path.